### PR TITLE
expression: rewrite builtin function TIMEDIFF

### PR DESCRIPTION
**Pull request by winkyao**
to #4080

### CODE REVIEW

**Comment by winkyao** [comment]
@XuHuaiyu @zz-jason @breeswish PTAL

File: expression/builtin_time.go [hunk 1]
```diff
@@ -336,79 +343,305 @@ func (c *timeDiffFunctionClass) getFunction(ctx context.Context, args []Expressi
...
 	}
-	sig := &builtinTimeDiffSig{newBaseBuiltinFunc(args, ctx)}
+
+	arg0Tp, arg1Tp := args[0].GetType().Tp, args[1].GetType().Tp
+	fsp := int(math.Max(float64(args[0].GetType().Decimal), float64(args[1].GetType().Decimal)))
+
+	var sig builtinFunc
+	// arg0 and arg1 must be the same time type(compatible), or timediff will return NULL.
+	// TODO: we don't really need Duration type, actually in MySQL, it use Time class to represent
+	// all the time type, and use type filed to distinguish datetime, date, timestamp or time(duration).
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: s/ type filed/ field type

File: expression/builtin_time.go [hunk 2]
```diff
@@ -336,79 +343,305 @@ func (c *timeDiffFunctionClass) getFunction(ctx context.Context, args []Expressi
...
+	// With the duration type, we are hard to port all the MySQL behavior.
+	switch arg0Tp {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]:
``` go
arg0Tp, arg1Tp := tpString, tpString

switch fieldType2EvalTp(arg0Tp){
case tpDuration:
    arg0Tp = tpDuration
case tpDatetime, tpTimestamp:
    arg0Tp = tpDatetime
}

switch fieldType2EvalTp(arg1Tp){
case tpDuration:
    arg1Tp = tpDuration
case tpDatetime, tpTimestamp:
    arg1Tp = tpDatetime
}

bf := newBaseBuiltinFuncWithTp(args, ctx, tpDuration, arg0Tp, arg1Tp)
bf.tp.Decimal = fsp

switch arg0Tp{
case tpDatetime:
    switch arg1Tp{
        case tpDatetime:
            sig = xxx
        case tpDuration:
        case tpString:
    }
case tpDuration:
case tpString:
}
```

File: util/types/time.go [hunk 3]
```diff
@@ -1033,23 +1032,19 @@ func ParseDuration(str string, fsp int) (Duration, error) {
...
-	}
-
+	d, err = TruncateOverflowMySQLTime(d)
 	return Duration{Duration: d, Fsp: fsp}, errors.Trace(err)
 }
 
-// TruncateOverflowMySQLTime truncates d when it overflows, and return truncated duration.
-func TruncateOverflowMySQLTime(d gotime.Duration) (gotime.Duration, bool) {
+// TruncateOverflowMySQLTime truncates d when it overflows, and return ErrTruncatedWrongVal.
+func TruncateOverflowMySQLTime(d gotime.Duration) (gotime.Duration, error) {
```
Comment (XuHuaiyu) [inline util/types/time.go]: add test for this function.

File: expression/builtin_time.go [hunk 4]
```diff
@@ -336,79 +343,305 @@ func (c *timeDiffFunctionClass) getFunction(ctx context.Context, args []Expressi
...
+
+	t1, err := val1.ConvertToTime(mysql.TypeDuration)
 	if err != nil {
-		return d, errors.Trace(err)
+		return d, false, errors.Trace(err)
 	}
+	return calculTimeDiff(t0, t1, sc)
+}
+
+func calculTimeDiff(t0, t1 types.Time, sc *variable.StatementContext) (d types.Duration, isNull bool, err error) {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: s/calculTimeDiff/ calculateTimeDiff

Comment (XuHuaiyu) [reply expression/builtin_time.go]: add comment for this func.

File: expression/builtin_time.go [hunk 5]
```diff
@@ -336,79 +343,305 @@ func (c *timeDiffFunctionClass) getFunction(ctx context.Context, args []Expressi
...
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_timediff
+func (b *builtinNullTimeDiffSig) evalDuration(row []types.Datum) (d types.Duration, isNull bool, err error) {
+	return d, true, nil
+}
+
+func getStrFsp(strArg string, fsp int) int {
+	if n := strings.IndexByte(strArg, '.'); n >= 0 {
+		lenStrFsp := len(strArg[n+1:])
+		if lenStrFsp <= types.MaxFsp {
+			fsp = int(math.Max(float64(lenStrFsp), float64(fsp)))
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: mathutil.Max()

File: expression/builtin_time.go [hunk 6]
```diff
@@ -336,79 +343,305 @@ func (c *timeDiffFunctionClass) getFunction(ctx context.Context, args []Expressi
...
+	baseDurationBuiltinFunc
+}
+
+// evalDuration evals a builtinNullTimeDiffSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_timediff
+func (b *builtinNullTimeDiffSig) evalDuration(row []types.Datum) (d types.Duration, isNull bool, err error) {
+	return d, true, nil
+}
+
+func getStrFsp(strArg string, fsp int) int {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: why add this func?

File: expression/builtin_time.go [hunk 7]
```diff
@@ -336,79 +343,305 @@ func (c *timeDiffFunctionClass) getFunction(ctx context.Context, args []Expressi
...
+	if n := strings.IndexByte(strArg, '.'); n >= 0 {
+		lenStrFsp := len(strArg[n+1:])
+		if lenStrFsp <= types.MaxFsp {
+			fsp = int(math.Max(float64(lenStrFsp), float64(fsp)))
+		}
+	}
+	return fsp
+}
+
+func convertStringToTime(str string, fsp int, sc *variable.StatementContext) (types.Time, error) {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: toTime or toDuration?
add comment for this function.

File: expression/builtin_time.go [hunk 8]
```diff
@@ -336,79 +343,305 @@ func (c *timeDiffFunctionClass) getFunction(ctx context.Context, args []Expressi
 	if err := c.verifyArgs(args); err != nil {
 		return nil, errors.Trace(err)
 	}
-	sig := &builtinTimeDiffSig{newBaseBuiltinFunc(args, ctx)}
+
+	arg0Tp, arg1Tp := args[0].GetType().Tp, args[1].GetType().Tp
+	fsp := int(math.Max(float64(args[0].GetType().Decimal), float64(args[1].GetType().Decimal)))
+
+	var sig builtinFunc
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: move the definition of sig and err to line 342.

File: expression/builtin_time.go (hunk 6 above)
Comment (winkyao) [reply expression/builtin_time.go]: the func exist before rewrite. used to find out correct fsp in duration string.

File: util/types/time.go (hunk 3 above)
Comment (winkyao) [reply util/types/time.go]: done!

**Comment by winkyao** [comment]
@XuHuaiyu PTAL

File: expression/builtin_time.go [hunk 9]
```diff
@@ -332,83 +339,297 @@ type timeDiffFunctionClass struct {
 	baseFunctionClass
 }
 
+func (c *timeDiffFunctionClass) getArgEvalTp(fieldTp *types.FieldType) evalTp {
+	argTp := tpString
+	tp := fieldTp2EvalTp(fieldTp)
+	switch tp {
```
Comment (zz-jason) [inline expression/builtin_time.go]:
```go
switch tp := fieldTp2EvalTp(fieldTp); tp {
    ...
}
```

File: expression/builtin_time.go [hunk 10]
```diff
@@ -332,83 +339,297 @@ type timeDiffFunctionClass struct {
...
+	d, err = val0.Sub(val1)
+	if err != nil {
+		return d, false, errors.Trace(err)
+	}
+
+	d.Duration, err = types.TruncateOverflowMySQLTime(d.Duration)
+	if types.ErrTruncatedWrongVal.Equal(err) {
+		err = sc.HandleTruncate(err)
+	}
+	return d, false, errors.Trace(err)
```
Comment (zz-jason) [inline expression/builtin_time.go]:
```go
return d, err != nil, errors.Trace(err)
```

File: expression/builtin_time.go [hunk 11]
```diff
@@ -332,83 +339,297 @@ type timeDiffFunctionClass struct {
...
-		err = types.ErrTruncatedWrongVal.GenByArgs("time", t.String())
+	d = t0.Sub(&t1)
+	d.Duration, err = types.TruncateOverflowMySQLTime(d.Duration)
+	if types.ErrTruncatedWrongVal.Equal(err) {
 		err = sc.HandleTruncate(err)
 	}
-	t.Duration = ret
-	d.SetMysqlDuration(t)
-	return
+	return d, false, errors.Trace(err)
```
Comment (zz-jason) [inline expression/builtin_time.go]:
```go
return d, err != nil, errors.Trace(err)
```

File: expression/builtin_time.go [hunk 12]
```diff
@@ -332,83 +339,297 @@ type timeDiffFunctionClass struct {
...
+		return d, isNull, errors.Trace(err)
 	}
-	return t, errors.Trace(err)
+
+	val1, isNull, err := b.args[1].EvalTime(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
+	}
+
+	return calculateTimeDiff(val0, val1, sc)
```
Comment (zz-jason) [inline expression/builtin_time.go]:
```go
result, isNull, err := calculateTimeDiff(val0, val1, sc)
return result, isNull, errors.Trace(err)
```

File: expression/builtin_time.go [hunk 13]
```diff
@@ -332,83 +339,297 @@ type timeDiffFunctionClass struct {
...
-		t, err = convertDatumToTime(sc, arg)
+type builtinTimeTimeTimeDiffSig struct {
+	baseDurationBuiltinFunc
+}
+
+// evalDuration evals a builtinTimeTimeTimeDiffSig.
+// See https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_timediff
+func (b *builtinTimeTimeTimeDiffSig) evalDuration(row []types.Datum) (d types.Duration, isNull bool, err error) {
+	sc := b.getCtx().GetSessionVars().StmtCtx
+	val0, isNull, err := b.args[0].EvalTime(row, sc)
```
Comment (zz-jason) [inline expression/builtin_time.go]: how about change `val0` to `lhs`, `val1` to `rhs` ?

File: expression/builtin_time.go [hunk 14]
```diff
@@ -332,83 +339,297 @@ type timeDiffFunctionClass struct {
...
+
+	val1, isNull, err := b.args[1].EvalDuration(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
+	}
+
+	t0, err := convertStringToDuration(val0, b.tp.Decimal, sc)
 	if err != nil {
-		return d, errors.Trace(err)
+		return d, false, errors.Trace(err)
```
Comment (zz-jason) [inline expression/builtin_time.go]:
```go
return d, true, errors.Trace(err)
```

File: expression/builtin_time.go (hunk 13 above)
Comment (winkyao) [reply expression/builtin_time.go]: lhs means `left-hand size`? I think it is not suit in here.

**Comment by winkyao** [comment]
@zz-jason @XuHuaiyu PTAL

File: expression/builtin_time.go (hunk 13 above)
Comment (zz-jason) [reply expression/builtin_time.go]: it is widely used in two-parameter functions

File: expression/builtin_time.go [hunk 15]
```diff
@@ -332,83 +339,303 @@ type timeDiffFunctionClass struct {
...
+
+	val1, isNull, err := b.args[1].EvalString(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
+	}
+
+	fsp := b.tp.Decimal
+	t0, err := convertStringToDuration(val0, fsp, sc)
+	if err != nil {
+		return d, false, errors.Trace(err)
```
Comment (zz-jason) [inline expression/builtin_time.go]:
```go
return d, true, errors.Trace(err)
```

File: expression/builtin_time.go [hunk 16]
```diff
@@ -332,83 +339,303 @@ type timeDiffFunctionClass struct {
...
+	}
+
+	t1, err := convertStringToDuration(val1, fsp, sc)
+	if err != nil {
+		return d, false, errors.Trace(err)
```
Comment (zz-jason) [inline expression/builtin_time.go]: ditto

**Comment by winkyao** [comment]
@zz-jason @XuHuaiyu PTAL

File: expression/builtin_time.go [hunk 17]
```diff
@@ -332,83 +339,319 @@ type timeDiffFunctionClass struct {
...
+	rhsStr, isNull, err := b.args[1].EvalString(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
 	}
 
-	sc := b.ctx.GetSessionVars().StmtCtx
-	fsp := int(math.Max(float64(args[0].Frac()), float64(args[1].Frac())))
-	t0, err := b.convertArgToTime(sc, args[0], fsp)
+	rhs, _, isDuration, err := convertStringToDuration(rhsStr, b.tp.Decimal, sc)
 	if err != nil {
```
Comment (zz-jason) [inline expression/builtin_time.go]:
```go
if err != nil || !isDuration {
    return d, true, errors.Trace(err)
}
```

File: expression/builtin_time.go [hunk 18]
```diff
@@ -332,83 +339,319 @@ type timeDiffFunctionClass struct {
...
+	}
+
+	rhs, isNull, err := b.args[1].EvalDuration(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
 	}
-	t1, err := b.convertArgToTime(sc, args[1], fsp)
+
+	lhs, _, isDuration, err := convertStringToDuration(lhsStr, b.tp.Decimal, sc)
 	if err != nil {
```
Comment (zz-jason) [inline expression/builtin_time.go]: ditto

File: expression/builtin_time.go [hunk 19]
```diff
@@ -332,83 +339,319 @@ type timeDiffFunctionClass struct {
...
 
-	t := t0.Sub(&t1)
-	ret, truncated := types.TruncateOverflowMySQLTime(t.Duration)
-	if truncated {
-		err = types.ErrTruncatedWrongVal.GenByArgs("time", t.String())
+	d, isNull, err = calculateDurationTimeDiff(lhs, rhs, sc)
+	return d, isNull, errors.Trace(err)
+}
+
+func calculateTimeDiff(t0, t1 types.Time, sc *variable.StatementContext) (d types.Duration, isNull bool, err error) {
```
Comment (zz-jason) [inline expression/builtin_time.go]: - move `sc *variable.StatementContext` to the first parameter 
- `s/t0/lhs/` and `s/t1/rhs/`

File: expression/builtin_time.go [hunk 20]
```diff
@@ -332,83 +339,319 @@ type timeDiffFunctionClass struct {
...
+	if types.ErrTruncatedWrongVal.Equal(err) {
 		err = sc.HandleTruncate(err)
 	}
-	t.Duration = ret
-	d.SetMysqlDuration(t)
-	return
+	return d, err != nil, errors.Trace(err)
+}
+
+func calculateDurationTimeDiff(t0, t1 types.Duration, sc *variable.StatementContext) (d types.Duration, isNull bool, err error) {
```
Comment (zz-jason) [inline expression/builtin_time.go]: ditto

File: expression/builtin_time.go [hunk 21]
```diff
@@ -332,83 +339,319 @@ type timeDiffFunctionClass struct {
...
+		return d, isNull, errors.Trace(err)
+	}
+
+	rhsStr, isNull, err := b.args[1].EvalString(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
+	}
+
+	_, rhs, isDuration, err := convertStringToDuration(rhsStr, b.tp.Decimal, sc)
+	if err != nil {
```
Comment (zz-jason) [inline expression/builtin_time.go]: ditto

File: expression/builtin_time.go [hunk 22]
```diff
@@ -332,83 +339,319 @@ type timeDiffFunctionClass struct {
...
+		return d, isNull, errors.Trace(err)
+	}
+
+	rhs, isNull, err := b.args[1].EvalTime(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
+	}
+
+	_, lhs, isDuration, err := convertStringToDuration(lhsStr, b.tp.Decimal, sc)
+	if err != nil {
```
Comment (zz-jason) [inline expression/builtin_time.go]: ditto

File: expression/builtin_time.go [hunk 23]
```diff
@@ -332,83 +339,319 @@ type timeDiffFunctionClass struct {
...
+		return d, isNull, errors.Trace(err)
+	}
+
+	rhs, isNull, err := b.args[1].EvalString(row, sc)
+	if isNull || err != nil {
+		return d, isNull, errors.Trace(err)
+	}
+
+	fsp := b.tp.Decimal
+	d0, t0, isDuration0, err := convertStringToDuration(lhs, fsp, sc)
```
Comment (zz-jason) [inline expression/builtin_time.go]: how about `s/d0/lhsDuration/`, `s/t0/lhsTime/` and `s/isDuration0/lhsIsDuration/` ?

File: expression/builtin_time.go [hunk 24]
```diff
@@ -332,83 +339,319 @@ type timeDiffFunctionClass struct {
...
+		if lenStrFsp <= types.MaxFsp {
+			fsp = mathutil.Max(lenStrFsp, fsp)
+		}
+	}
+	return fsp
+}
+
+// convertStringToDuration converts string to duration, it return types.Time because in some case
+// it will converts string to datetime.
+func convertStringToDuration(str string, fsp int, sc *variable.StatementContext) (d types.Duration, t types.Time,
```
Comment (zz-jason) [inline expression/builtin_time.go]: move `sc *variable.StatementContext` to the first parameter

File: expression/builtin_time.go [hunk 25]
```diff
@@ -332,83 +339,319 @@ type timeDiffFunctionClass struct {
...
+	isDuration bool, err error) {
+	fsp = getStrFsp(str, fsp)
```
Comment (zz-jason) [inline expression/builtin_time.go]: is function `getStrFsp()` only used here ? if so, maybe we can eliminate that function ?

File: plan/join_reorder.go [hunk 26]
```diff
@@ -47,7 +47,7 @@ func findColumnIndexByGroup(groups []LogicalPlan, col *expression.Column) int {
 			return i
 		}
 	}
-	log.Errorf("Unknown columns %s, from id %d, position %d", col, col.FromID, col.Position)
+	log.Errorf("Unknown columns %s, from id %v, position %d", col, col.FromID, col.Position)
```
Comment (zz-jason) [inline plan/join_reorder.go]: why change "%d" to "%v" ?

Comment (winkyao) [reply plan/join_reorder.go]: my local go vet will report "plan/join_reorder.go:50: arg col.FromID for printf verb %d of wrong type: string"

File: util/types/convert.go [hunk 27]
```diff
@@ -171,9 +171,11 @@ func StrToDateTime(str string, fsp int) (Time, error) {
 	return ParseTime(str, mysql.TypeDatetime, fsp)
 }
 
-// StrToDuration converts str to Duration.
+// StrToDuration converts str to Duration. It returns Duration in normal case,
+// and returns Time when str is in datetime format.
+// when isDuration is true, the d is returned, when it is false, the t is returned.
 // See https://dev.mysql.com/doc/refman/5.5/en/date-and-time-literals.html.
-func StrToDuration(sc *variable.StatementContext, str string, fsp int) (t Time, err error) {
+func StrToDuration(sc *variable.StatementContext, str string, fsp int) (d Duration, t Time, isDuration bool, err error) {
```
Comment (XuHuaiyu) [inline util/types/convert.go]: Str2TimeOrDuration better?

File: expression/builtin_time.go [hunk 28]
```diff
@@ -332,83 +339,299 @@ type timeDiffFunctionClass struct {
...
+	lhs, _, isDuration, err := convertStringToDuration(sc, lhsStr, b.tp.Decimal)
+	if err != nil || !isDuration {
+		return d, true, errors.Trace(err)
+	}
+
+	d, isNull, err = calculateDurationTimeDiff(sc, lhs, rhs)
+	return d, isNull, errors.Trace(err)
+}
+
+func calculateTimeDiff(sc *variable.StatementContext, lhs, rhs types.Time) (d types.Duration, isNull bool, err error) {
```
Comment (XuHuaiyu) [inline expression/builtin_time.go]: add comments for `calculateTimeDiff ` and `calculateDurationTimeDiff `

File: util/types/convert.go (hunk 27 above)
Comment (winkyao) [reply util/types/convert.go]: MySQL use strToTime, so I think no need to change it...the better way is to eliminate the duration type totally.

**Comment by winkyao** [comment]
@XuHuaiyu PTAL

**Comment by XuHuaiyu** [comment]
LGTM

**Pull request merged**